File: Laboratory/Lab1/scala/list1.scala
// import scala.annotation.tailrec

def reverse4[A, B, C, D](list: (A, B, C, D)): (D, C, B, A) = {
  (list._4, list._3, list._2, list._1)
}

// def sumProd(s: Int, e: Int): (Int, Int) = {
//   if s==e then return (0, 0)
//   def createList(s: Int, e: Int): List[Int] = {
//     if s >= e then Nil
//     else s :: createList(s + 1, e)
//   }

//   val list = createList(s, e)
//   val (sum, prod) = (0, 1)

//   def count(list: List[Int]): (Int, Int) = {
//     if list == Nil then (sum, prod)
//     else {
//       val (sum, prod) = count(list.tail)
//       (sum + list.head, prod * list.head)
//     }
//   }

//   count(list)
// }

def sumProd(s: Int, e: Int): (Int, Int) = {
  if s>=e then (0, 1)
  else {
    val (sum, prod) = sumProd(s + 1, e)
    (s + sum, s * prod)
  }
}

// def isPerfectTailRec(n: Int): Boolean = {
//   @tailrec
//   def divider(sum: Int, i: Int): Int = {
//     if i == 0 then sum
//     else if n % i == 0 then divider(sum + i, i - 1)
//     else divider(sum, i - 1)
//   }

//   n == divider(0, n - 1)
// }

def isPerfect(n: Int): Boolean = {
  if n < 1 then return false

  def dividerList(dividers: List[Int], i: Int): List[Int] = {
    if i < n then if n % i == 0 then i :: dividerList(dividers, i + 1) else dividerList(dividers, i + 1)
    else dividers
  }

  def sum(list: List[Int]): Int = {
    if list == Nil then 0
    else list.head + sum(list.tail)
  }

  sum(dividerList(Nil, 1)) == n
}

def insert[A](list: List[A], element: A, index: Int): List[A] = {
  if list == Nil then List(element)
  else {
    if index == 0 then element :: list
    else list.head :: insert(list.tail, element, index - 1)
  }
}

println("Test for reverse4")
println(reverse4(true, 'd', "string", 1) == (1, "string", 'd', true))
println(reverse4(1, 2, 3, 4) == (4, 3, 2, 1))

println("Test for sumProd")
println(sumProd(1, 5) == (10, 24))
println(sumProd(1, 1) == (0, 1))
println(sumProd(-1, -10) == (0, 1))
println(sumProd(1, 0) == (0, 1))
println(sumProd(1, 2) == (1, 1))
println(sumProd(-10, -5) == (-40, -30240))
println(sumProd(0, 0)==(0, 1))


println("Test for isPerfect")
val list = (0 to 500).toList
val perfectNumbers = list.filter(isPerfect)


println("Test for insert")
println(insert(List(1, 2, 3), 4, 0) == List(4, 1, 2, 3))
println(insert(List(1, 2, 3), 4, 1) == List(1, 4, 2, 3))
println(insert(List('h', 'e', 'l', 'o'), 'l', 2) == List('h', 'e', 'l', 'l', 'o'))
println(insert(List(true, false, true), false, -1) == List(true, false, true, false))
println(insert(List('1', '2', '3'), '4', 4) == List('1', '2', '3', '4'))

// modyfikacja 
def choice(list1: List[Int], list2: List[Int]): List[Int] = {
  if list1==Nil && list2==Nil then Nil
  else if list1 == Nil && list2!=Nil then list2
  else if list1 != Nil && list2==Nil then list1
  else if list1.head > list2.head then list1.head :: choice(list1.tail, list2.tail)
  else list2.head :: choice(list1.tail, list2.tail)
  // else if list1.head < list2.head then list2.head :: choice(list1.tail, list2.tail)
  // else list1.head :: choice(list1.tail, list2.tail)
}

println("Test for choice")
choice(List(1,2,3), List(4,5,6))==List(4,5,6)
choice(List(1,2,3), List(1,2,3))==List(1,2,3)
choice(List(1,2,3), List(0,0,0))==List(1,2,3)
choice(List(1,2,3), List(0,1,2))==List(1,2,3)
choice(List(1, 5, 3 ,8), List(2, 7, 4, 1))==List(2, 7, 4, 8)
choice(List(1, 5, 3 ,8), List(2, 7, 4, 1, 9))==List(2, 7, 4, 8, 9)
choice(List(1, 5, 3 ,8, 9), List(2, 7, 4, 1))==List(2, 7, 4, 8, 9)
choice(List(0,0,0,0), List()) == List(0,0,0,0)

File: Laboratory/Lab1/ocaml/list1.ml
let reverse4 (a, b, c, d) = (d, c, b, a)

(* let sumProd s e =
  let rec createList s e =
    if s >= e then []
    else s :: (createList (s + 1) e)
  in

  let list = createList s e in
  let sum, prod = 0, 1 in

  let rec count list sum prod =
    if list = [] then (sum, prod)
    else count (List.tl list) (sum + List.hd list) (prod * List.hd list)
  in

  count list sum prod;; *)

let rec sumProd s e =
  if s >= e then (0, 1)
  else
    let sum, prod = sumProd (s + 1) e in
    (s + sum, s * prod)

let isPerfect n = 

  let rec dividersList dividers i n =
    if i < n then if n mod i = 0 then i :: dividersList dividers (i + 1) n else dividersList dividers (i + 1) n
    else dividers
  in
    
  let rec letsSum list =
    if list = [] then 0
    else List.hd list + letsSum (List.tl list)
  in
  
  if n < 1 then false
  else 
    let resSum = letsSum (dividersList [] 1 n) in 
    if resSum = n then true else false;;

let rec insert list element index =
  if list = [] then [element]
  else if index = 0 then element :: list
  else List.hd list :: insert (List.tl list) element (index - 1);;


(* Testy dla reverse4 *)
(reverse4 (1, 2, 3, 4) = (4, 3, 2, 1));;
(reverse4 ('1', '1', '1', '1') = ('1', '1', '1', '1'));;
(reverse4 (true, 1, true, "string") = ("string", true, 1, true));;

(* Testy dla sumProd *)
(sumProd 1 5) = (10, 24);;
(sumProd 1 1) = (0, 1);;
(sumProd (-1) (-10)) = (0, 1);;
(sumProd 1 0) = (0, 1);;
(sumProd 1 2) = (1, 1);;
(sumProd (-10) (-5)) = (-40, -30240);;

(* Testy dla isPerfect *)
(isPerfect 6 = true);;
(isPerfect 28 = true);;
(isPerfect 496 = true);;
(isPerfect 0 = false);;
(isPerfect 1 = false);;

(* Testy dla insert *)
(insert [1; 2; 3] 4 0 = [4; 1; 2; 3]);;
(insert [1; 2; 3] 4 1 = [1; 4; 2; 3]);;
(insert ['h'; 'e'; 'l'; 'o'] 'l' 2 = ['h'; 'e'; 'l'; 'l'; 'o']);;
(insert [true; false; true] false (-1) = [true; false; true; false]);;
(insert ['1'; '2'; '3'] '4' 4 = ['1'; '2'; '3'; '4']);;


let rec choice list1 list2 = 
  if list1 = [] && list2 = [] then [] 
  else if list1 = [] && list2 != [] then list2
  else if list1 != [] && list2 = [] then list1
  else if List.hd list1 > List.hd list2 then List.hd list1 :: choice (List.tl list1)  (List.tl list2)
  else List.hd list2 :: choice (List.tl list1) (List.tl list2);;
  (* else if List.hd list1 < List.hd list2 then List.hd list2 :: choice (List.tl list1)  (List.tl list2) *)
  (* else List.hd list1 :: choice (List.tl list1) (List.tl list2);; *)


(* Test for choice *)
(choice [1; 2; 3] [4; 5; 6] = [4; 5; 6]);;
(choice [1; 2; 3] [1; 2; 3] = [1; 2; 3]);;
(choice [1; 2; 3] [1; 2; 3; 4] = [1; 2; 3; 4]);;
(choice [1; 2; 3; 4] [1; 2; 3] = [1; 2; 3; 4]);;
(choice [1; 5; 3; 8] [2; 7; 4; 1] = [2; 7; 4; 8]);;
(choice [1; 2; 3; 4] [] = [1; 2; 3; 4]);;

